router/h2/src/main/scala/io/buoyant/router/h2/StreamStatsFilter.scala

```diff
@@ -1,64 +1,104 @@
-      override def role: Stack.Role = StreamStatsFilter.role
-      override def description = "Record stats on h2 streams"
+    new Stack.Module4[param.Stats, h2param.H2StreamClassifier, param.ExceptionStatsHandler, Param, ServiceFactory[Request, Response]] {
+      override val role: Stack.Role = StreamStatsFilter.role
+      override val description = "Record stats on h2 streams"
       override def make(
         statsP: param.Stats,
         classifierP: h2param.H2StreamClassifier,
+        handlerP: param.ExceptionStatsHandler,
+        timeP: Param,
         next: ServiceFactory[Request, Response]
       ): ServiceFactory[Request, Response] = {
         val param.Stats(stats) = statsP
         val h2param.H2StreamClassifier(classifier) = classifierP
-        new StreamStatsFilter(stats, classifier).andThen(next)
+        val param.ExceptionStatsHandler(handler) = handlerP
+        val Param(timeUnit) = timeP
+        new StreamStatsFilter(stats, classifier, handler, timeUnit).andThen(next)
       }
     }
+
+  private[StreamStatsFilter] val SyntheticException =
+    syntheticException
+
 }
 
-class StreamStatsFilter(statsReceiver: StatsReceiver, classifier: H2StreamClassifier)
-  extends SimpleFilter[Request, Response] {
+class StreamStatsFilter(
+  statsReceiver: StatsReceiver,
+  classifier: H2StreamClassifier,
+  exceptionStats: ExceptionStatsHandler,
+  timeUnit: TimeUnit
+) extends SimpleFilter[Request, Response] {
+
+  private[this] val latencyStatSuffix: String =
+    timeUnit match {
+      case TimeUnit.NANOSECONDS => "ns"
+      case TimeUnit.MICROSECONDS => "us"
+      case TimeUnit.MILLISECONDS => "ms"
+      case TimeUnit.SECONDS => "secs"
+      case _ => timeUnit.toString.toLowerCase
+    }
 
   class StreamStats(
     protected val stats: StatsReceiver,
     durationName: Option[String] = None
   ) {
-    private[this] val durationMs =
-      stats.stat(s"${durationName.getOrElse("stream_duration")}_ms")
+    private[this] val duration =
+      stats.stat(s"${durationName.getOrElse("stream_duration")}_$latencyStatSuffix")
     private[this] val successes = stats.counter("stream_success")
     private[this] val failures = stats.counter("stream_failures")
 
     def success(streamDuration: Duration): Unit = {
-      durationMs.add(streamDuration.inMillis)
+      duration.add(streamDuration.inUnit(timeUnit))
       successes.incr()
     }
 
-    def failure(streamDuration: Duration): Unit = {
-      durationMs.add(streamDuration.inMillis)
-      successes.incr()
+    def failure(streamDuration: Duration, e: Throwable): Unit = {
+      duration.add(streamDuration.inUnit(timeUnit))
+      failures.incr()
+      exceptionStats.record(stats, e)
     }
 
     private[this] val frameBytes = stats.stat("data_bytes")
 
     def apply(
       startT: Stopwatch.Elapsed,
-      onFinalFrame: Option[Try[Frame]] => Unit = { _ => }
+      onFinalFrame: Option[Try[Frame]] => Try[Unit] => Unit = { _ => _ => }
```

Comment:
I don't think we want to pass an addition `Try[Unit]` in here to capture the potential exception.  Even if we did, `Option[Throwable]` is probably a better type than `Try[Unit]`.  See below for suggestions on how to eliminate this.